plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/lambda.ts

```diff
@@ -0,0 +1,119 @@
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Span, SpanKind, Tracer } from '@opentelemetry/api';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import {
+  AwsSdkInstrumentationConfig,
+  NormalizedRequest,
+  NormalizedResponse,
+} from '../types';
+import { RequestMetadata, ServiceExtension } from './ServiceExtension';
+import { TextMapSetter, context, propagation } from '@opentelemetry/api';
+import type { ClientContext } from 'aws-lambda';
+
+export class LambdaServiceExtension implements ServiceExtension {
+  requestPreSpanHook(request: NormalizedRequest): RequestMetadata {
+    const functionName = this.extractFunctionName(request.commandInput);
+
+    let spanAttributes = {};
+    let spanName: string | undefined;
+
+    switch (request.commandName) {
+      case 'Invoke':
+        spanAttributes = {
+          [SemanticAttributes.FAAS_INVOKED_NAME]: functionName,
+          [SemanticAttributes.FAAS_INVOKED_PROVIDER]: 'aws',
+        };
+        if (request.region) {
+          spanAttributes = {
+            ...spanAttributes,
+            [SemanticAttributes.FAAS_INVOKED_REGION]: request.region,
+          };
+        }
+        spanName = `${functionName} invoke`;
+        break;
+    }
+    return {
+      isIncoming: false,
+      spanAttributes,
+      spanKind: SpanKind.CLIENT,
+      spanName,
+    };
+  }
+
+  requestPostSpanHook = (request: NormalizedRequest) => {
+    switch (request.commandName) {
+      case 'Invoke':
+        {
+          request.commandInput = injectPropagationContext(request.commandInput);
+        }
+        break;
+    }
+  };
+
+  responseHook(
+    response: NormalizedResponse,
+    span: Span,
+    tracer: Tracer,
+    config: AwsSdkInstrumentationConfig
+  ) {
+    const operation = response.request.commandName;
+
+    if (operation === 'Invoke') {
+      if (response.data && '$metadata' in response.data) {
+        span.setAttribute(
+          SemanticAttributes.FAAS_EXECUTION,
+          response.data['$metadata'].requestId
+        );
+      }
+    }
+  }
+
+  extractFunctionName = (commandInput: Record<string, any>): string => {
+    return commandInput?.FunctionName;
+  };
+}
+
+class ContextSetter implements TextMapSetter<Record<string, any>> {
+  set(carrier: Record<string, any>, key: string, value: string): void {
+    const parsedClientContext: ClientContext = JSON.parse(
+      carrier.ClientContext !== undefined
+        ? Buffer.from(carrier.ClientContext, 'base64').toString('utf8')
+        : '{"Custom":{}}'
+    );
+    const updatedPayload = {
+      ...parsedClientContext,
+      Custom: {
```

Comment:
Where is the code in `@opentelemetry/instrumentation-aws-lambda` that extracts the context from this `Custom` field?

Reply:
The `eventContextExtractor` configuration parameter to `AwsLambdaInstrumentation ` in `@opentelemetry/instrumentation-aws-lambda` allows this to be extracted from context. There's an example implementation of this in the tests in [`@opentelemetry/instrumentation-aws-lambda`](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/760c4d4e9bd500ab65980f7414f72ad93da42e2f/plugins/node/opentelemetry-instrumentation-aws-lambda/test/integrations/lambda-handler.test.ts#L660)

Reply:
So if the user will install the `aws-sdk` and `aws-lambda` instrumentations, he will not get automatic context propagation, but will have to add specific code for that, right?
I think we could add the extraction to the `aws-lambda` [_defaultEventContextExtractor](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/986c3499b324ebdf49aabf35bc3711c91bb91ec8/plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts#L327) to give users a better experience. Not required for this PR, we can open an issue to add it and someone will do it in the future (you are welcome to do that yourself if you have motivation and time :) )

BTW, do you know maybe if this `Custom` field is something standard in AWS? Is it used in other contexts or was introduced only for OTEL? Do we know if instrumentations in other languages use it as well? @NathanielRN @willarmiros

Reply:
I'll add it to my list of PRs integrating honeycomb tracing via OTEL has spawned 😄 

I believe `Custom` to be standard - it's not wonderfully documented from the invoker's perspective, but is mentioned in the [node developer guide for lambda](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html), and I've done some cross-language checks and e.g. Python also supports it.

Reply:
awsome! thanks for the research and for the willingness to fix that in `aws-lambda` instrumentation as well :)